### PR TITLE
Comprehensive ML Pipeline Upgrade: Ensemble Support, CNN/ResBlock Models, and Improved Docs

### PR DESCRIPTION
**Pull request by JannesB24**
This pull request introduces significant updates to the project, focusing on improving the machine learning pipeline, adding new models, and enhancing documentation and setup processes. The most important changes include the addition of ensemble learning functionality, new model implementations, updates to the data loading pipeline, and improvements to project setup and documentation.

### Machine Learning Pipeline Enhancements:

* **Ensemble Learning Implementation**: Added a new file `udl_project/ensemble_addon.py` to implement ensemble training and comparison functionality. This includes training multiple ResBlock models, comparing their performance with the original model, and analyzing overfitting reduction.

* **New Models**:
  - [`udl_project/models/cnn.py`](diffhunk://#diff-1b9c3e65c6e10ee83d4a5992795f8c06da81c9b8f1a949b7699c830c705ff5a7R1-R27): Introduced a CNN model with feature extraction and classification layers.
  - [`udl_project/models/res_block.py`](diffhunk://#diff-eb1791bacfed6d8f70f40de47f5eb53b00c0dc76d8a6e8dd7d9b728491529039R1-R51): Added a ResBlock model with bottleneck layers and residual connections.
  - [`udl_project/models/res_net_cnn.py`](diffhunk://#diff-645171720a43ca3e5d0a741b7f27501e165c1ce9878f9690838780402fd85649R1-R19): Added a placeholder for ResNetCNN implementation.

### Data Loading Pipeline Updates:

* **Image DataLoader for Flower Dataset**: Added `udl_project/DataloaderFlowerSet.py`, which loads training, validation, and test datasets with transformations like resizing, normalization, and augmentation for flower classification.

* **General DataLoader**: Added `udl_project/DataLoaderFFSet.py` to load image datasets with simple transformations and split them into training and testing sets.

### Project Setup and Documentation Improvements:

* **Updated `README.md`**:
  - Added a project directory tree view and instructions for generating it.
  - Included an alternative command (`make run-pc`) for running pre-commit hooks.

* **Dependency Management**:
  - Added new dependencies (`torch`, `torchvision`, `matplotlib`) to `infrastructure/requirements.in`.
  - Updated `infrastructure/requirements.txt` with pinned versions of dependencies, including those required for PyTorch-based workflows.

### Code Cleanup:

* **Removed `main.py`**: Deleted a standalone script for computing the mean of a NumPy array, as it is no longer relevant to the project scope.

### Pre-commit Hook Adjustments:

* **Modified `.pre-commit-config.yaml`**:
  - Updated arguments for `ruff-check` and `ruff-format` hooks to include the current directory (`.`).

* **Updated `Makefile`**:
  - Changed the `run-pc` target to use `--all-files` for running pre-commit hooks.

### CODE REVIEW

File: udl_project/DataloaderFlowerSet.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+train_loader = torch.utils.data.DataLoader(train_ds, batch_size=batch_size, shuffle=True)
+
+
+val_ds = torchvision.datasets.ImageFolder(val_dir, transform=transform)
+
+val_loader = torch.utils.data.DataLoader(val_ds, batch_size=batch_size, shuffle=True)
+
+
+test_ds = torchvision.datasets.ImageFolder(train_dir, transform=transform)
```
[the remark this answers is not in the document]
Comment (JannesB24) [reply udl_project/DataloaderFlowerSet.py:33]: Data Loader not used anymore remove later, right @FlatCatSociety ?

File: udl_project/DataLoaderFFSet.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+# Create DataLoader's
+train_dataloader_simple = DataLoader(
+    train_data, batch_size=BATCH_SIZE, shuffle=True, num_workers=NUM_WORKERS
+)
+
+test_dataloader_simple = DataLoader(
+    test_data, batch_size=BATCH_SIZE, shuffle=False, num_workers=NUM_WORKERS
+)
+
+train_dataloader_simple, test_dataloader_simple
```
[the remark this answers is not in the document]
Comment (JannesB24) [reply udl_project/DataLoaderFFSet.py:54]: TODO:
